### ISSUE TITLE
WPAS Session Cookies Affecting Caching Mechanism

### ISSUE THREAD

**Issue by Laurel-Summit-Media** [Awesome-Support/Awesome-Support#345]
Awesome Support is creating a cookie which seems to be created for each visitor our website. Can you please have a look at this below and tell me why this is happening? It was my understanding that the newest version of WPAS dropped user sessions. This is affecting the performance of our caching mechanism on our website. 

**curl -sI s20217.p628.sites.pressdns.com | grep "Set-Cookie|X-Cache-Status"
Set-Cookie: _wpas_session=0566e4e49c1561082d21ec32872ded89%7C%7C1477055661%7C%7C1477055301; expires=Fri, 21-Oct-2016 13:14:21 GMT; Max-Age=1800; path=/
X-Cache-Status: BYPASS

curl -sI s20217.p628.sites.pressdns.com | grep "Set-Cookie|X-Cache-Status"
Set-Cookie: _wpas_session=6d33c4e699fdb02cc892158a9b41c512%7C%7C1477055662%7C%7C1477055302; expires=Fri, 21-Oct-2016 13:14:22 GMT; Max-Age=1800; path=/
X-Cache-Status: BYPASS

curl -sI s20217.p628.sites.pressdns.com | grep "Set-Cookie|X-Cache-Status"
Set-Cookie: _wpas_session=1106c4977ff4bf87dfc19900e05c7d3a%7C%7C1477055664%7C%7C1477055304; expires=Fri, 21-Oct-2016 13:14:24 GMT; Max-Age=1800; path=/
X-Cache-Status: BYPASS**

The above illustrates how the cache is then bypassed which significantly affects performance since the entire caching mechanism is uninvolved which causes loss of acceleration. The "wpas" cookie in question seems to originate from the "awesome-support" plugin - would it be possible for the plugin to only create the cookie for authenticated users instead of all traffic? 

Pagely has an article concerning the effects of cookies and sessions in terms of performance: https://support.pagely.com/hc/en-us/articles/209015317-Optimizing-Your-WordPress-Site-With-Pagely

**Comment by julien731** [comment]
There is indeed a cookie that's being set for each visitor. This is because we're using [WP Session Manager](https://github.com/ericmann/wp-session-manager) to handle fake sessions.

Sessions are necessary for the plugin to work properly. While we are no longer using actual PHP sessions ( see #276 ) we still need to use the *fake* session manager.

I think that if you're having caching issues, maybe we should have a look at the stack level. What is your server stack?

**Comment by nachumie** [comment]
Same issue, pagely provided this:
Two issues we are noticing is that the site is not utilizing our Nginx caching layer and the site is making a large amount of admin-ajax.php requests. Both of these points cause PHP to overwork since there is only so many requests the PHP workers can handle.

1.) For the Nginx caching, you need to remove PHP sessions from the homepage/subpages. The Cloudflare PHP session is fine, but the other sessions need to be removed from any pages you don't need a session_start function. For example, the homepage has two PHP sessions:

Set-Cookie: _wpas_session=6dba0ae4cbf528797070fd1f1b1c6545%7C%7C1492719196%7C%7C1492718836; expires=Thu, 20-Apr-2017 20:13:16 GMT; Max-Age=1800; path=/
Set-Cookie: PHPSESSID=dukl85kq2nnh1knknbhr1voc54; path=/
From this curl request on the homepage:

 curl -I https://www.sabrent.com
HTTP/1.1 200 OK
Date: Thu, 20 Apr 2017 19:43:17 GMT
Content-Type: text/html; charset=UTF-8
Connection: keep-alive
Set-Cookie: __cfduid=dd6077dd5915fa20d97b243adda788dee1492717396; expires=Fri, 20-Apr-18 19:43:16 GMT; path=/; domain=.sabrent.com; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate
Pragma: no-cache
Link: <https://www.sabrent.com/wp-json/>; rel="https://api.w.org/", <https://www.sabrent.com/>; rel=shortlink
Set-Cookie: _wpas_session=6dba0ae4cbf528797070fd1f1b1c6545%7C%7C1492719196%7C%7C1492718836; expires=Thu, 20-Apr-2017 20:13:16 GMT; Max-Age=1800; path=/
Set-Cookie: PHPSESSID=dukl85kq2nnh1knknbhr1voc54; path=/
Vary: Accept-Encoding, User-Agent
X-User-Agent: standard
X-Cache-Config: 0 0
X-Cache-Status: MISS
Strict-Transport-Security: max-age=0; includeSubDomains; preload
X-Content-Type-Options: nosniff
Server: cloudflare-nginx
CF-RAY: 352a976f582e77c0-LAX
2.) Cut down on the amount of admin-ajax.php requests being made by the site. Here is the current usage from April 18th - April 20th:

Header Status Returns: non-HIT's, sorted by most up top:
  76767 200-"POST-www.sabrent.com/wp-admin/admin-ajax.php
   4219 200-"GET-sabrent.admin.vps.pagelyhosting.com/check.php
   2732 200-"GET-www.sabrent.com/
   2731 200-"GET-www.sabrent.com/downloads/
   1264 200-"GET-www.sabrent.com/downloads/?wpdmc=windows-drivers
    995 200-"GET-www.sabrent.com/products/
    891 200-"POST-www.sabrent.com/?wc-ajax=get_refreshed_fragments
    882 200-"HEAD-www.sabrent.com/
This article goes over these points a bit more and is a great starting point to address these two issues:

https://support.pagely.com/hc/en-us/articles/209015317-Optimizing-Your-WordPress-Site-With-Pagely



We unfortunately can't remove sessions / cookies on your behalf because it might affect website functionality and should please be considered by your web developer instead. I have narrowed down the two plugins download-manager and awesome-support which are responsible for the creation of the sessions:

wp-content/plugins/download-manager/download-manager.php
wp-content/plugins/awesome-support/includes/class-wpas-session.php: define( 'WP_SESSION_COOKIE', '_wpas_session' );
These plugins should please be disabled or reworked by your web developer so that the sessions can be avoided.

**Comment by nschvabenitz** [comment]
Julian,

I'm sorry I never replied to your question. We were using Pagely as well. Just as @nachumie has said. WPAS brakes pretty much all caching on the server because of the sessions.

**Comment by glagonikas** [comment]
@nachumie appreciate this a late response, but have you tried removing the cookies on the nginx config?

Assuming you're using something similar to https://websiteforstudents.com/wordpress-fastcgi-caching-nginx-ubuntu-17-04-17-10/ you can remove them using
```
if ($skip_cache = 0 ) {
    more_clear_headers "Set-Cookie*";
    more_clear_headers "Pragma";
}
```
We use this code for both Awesome Support and WooCommerce and it appears to work without issues on our setup.

### Preventing cookies on the application level
A recent update of the plugin introduced the wpas_allow_soft_loading filter that you can use on your application to prevent AS from loading and therefore creating a cookie. We've made a mu-plugin to prevent AS loading for spiders and bots.
